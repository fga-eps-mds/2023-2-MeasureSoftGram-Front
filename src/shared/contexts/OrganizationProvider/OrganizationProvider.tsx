import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback } from 'react';

import { Organization } from '@customTypes/organization';
import { organizationQuery } from '@services/organization';
import { useRequest } from '@hooks/useRequest';
import { useRouter } from 'next/router';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization | undefined;
  setCurrentOrganization: (organization: Organization) => void;
  organizationList: Organization[];
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const router = useRouter();

  const [currentOrganization, setCurrentOrganization] = useState<Organization | undefined>();

  const { data } = useRequest<{ results: [Organization] }>(organizationQuery.getAllOrganization());

  const organizationList = useCallback(() => data?.results ?? [], [data])();

  const value = useMemo(() => {
    const regex = /\d+/g;
    const queryProduct = router.query?.product as string;
    const organizationIndex = regex.exec(queryProduct);
    if (!currentOrganization && organizationList.length > 0 && queryProduct && organizationIndex) {
      setCurrentOrganization(organizationList[parseInt(organizationIndex[0], 10) - 1]);
    }
    return { currentOrganization, setCurrentOrganization, organizationList };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization, organizationList]);

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationContext');
  }

  return context;
}
