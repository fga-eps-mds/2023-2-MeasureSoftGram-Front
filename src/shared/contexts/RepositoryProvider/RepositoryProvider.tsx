import React, { createContext, useState, useContext, ReactNode, useMemo } from 'react';

import { Repository, Historical } from '@customTypes/repository';

interface Props {
  children: ReactNode;
}

interface IRepositoryContext {
  currentRepository?: Repository;
  setCurrentRepository: (repository: Repository) => void;
  repositoryList?: Repository[];
  setRepositoryList: (repository: Repository) => void;
  characteristics: string[];
  setCharacteristics: (characteristics: string[]) => void;
  subCharacteristics: string[];
  setSubCharacteristics: (subCharacteristics: string[]) => void;
  measures: string[];
  setMeasures: (measures: string[]) => void;
  metrics: string[];
  setMetrics: (metrics: string[]) => void;
  historicalTSQMI: Historical;
  setHistoricalTSQMI: (historical: Historical) => void;
}

const RepositoryContext = createContext<IRepositoryContext | undefined>(undefined);

export function RepositoryProvider({ children }: Props) {
  const [currentRepository, setCurrentRepository] = useState<Repository | undefined>();
  const [repositoryList, setRepositoryList] = useState<Repository[]>();

  const [characteristics, setCharacteristics] = useState<string[]>([]);
  const [subCharacteristics, setSubCharacteristics] = useState<string[]>([]);
  const [measures, setMeasures] = useState<string[]>([]);
  const [metrics, setMetrics] = useState<string[]>([]);
  const [historicalTSQMI, setHistoricalTSQMI] = useState<Historical>();

  const value = useMemo(
    () => ({
      currentRepository,
      setCurrentRepository,
      repositoryList,
      setRepositoryList,
      characteristics,
      setCharacteristics,
      subCharacteristics,
      setSubCharacteristics,
      measures,
      setMeasures,
      metrics,
      setMetrics,
      historicalTSQMI,
      setHistoricalTSQMI
    }),
    [currentRepository, repositoryList, characteristics, subCharacteristics, measures, metrics, historicalTSQMI]
  );

  return <RepositoryContext.Provider value={value}>{children}</RepositoryContext.Provider>;
}

export function useRepositoryContext() {
  const context = useContext(RepositoryContext);

  if (context === undefined) {
    throw new Error('useRepositoryContext must be used within a RepositoryContext');
  }

  return context;
}
