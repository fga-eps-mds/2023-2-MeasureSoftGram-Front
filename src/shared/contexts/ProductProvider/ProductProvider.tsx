import React, { createContext, useState, useContext, ReactNode, useMemo, useCallback, useEffect } from 'react';

import { Product } from '@customTypes/product';
import { productQuery } from '@services/product';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

interface Props {
  children: ReactNode;
}

export interface IProductContext {
  currentProduct?: Product;
  setCurrentProduct: (product: Product) => void;
  productsList?: Product[];
  updateProductList: (products: Product[]) => void;
}

export const ProductContext = createContext<IProductContext | undefined>(undefined);

export function ProductProvider({ children }: Props) {
  const [currentProduct, setCurrentProduct] = useState<Product | undefined>();
  const [productsList, setProductsList] = useState<Product[]>([]);

  const { currentOrganization } = useOrganizationContext();

  const updateProductList = useCallback((products: Product[]) => {
    setProductsList(products);
  }, []);

  const loadAllProducts = async () => {
    try {
      if (!currentOrganization) {
        updateProductList([]);
        return;
      }

      const result = await productQuery.getAllProducts(currentOrganization.id);

      updateProductList(result.data.results);
    } catch (error) {
      console.error(error);
    }
  };

  useEffect(() => {
    if (currentOrganization) {
      loadAllProducts();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentOrganization]);

  const value = useMemo(
    () => ({
      currentProduct,
      setCurrentProduct,
      productsList,
      updateProductList
    }),
    [currentProduct, productsList, updateProductList]
  );

  return <ProductContext.Provider value={value}>{children}</ProductContext.Provider>;
}

export function useProductContext() {
  const context = useContext(ProductContext);

  if (!context) {
    throw new Error('useProductContext must be used within a ProductContext');
  }

  return context;
}