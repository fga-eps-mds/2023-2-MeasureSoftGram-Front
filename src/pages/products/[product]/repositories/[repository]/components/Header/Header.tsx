import { useRepositoryContext } from '@contexts/RepositoryProvider';
import { Box, Typography } from '@mui/material';

import React from 'react';

function Header() {
  const { currentRepository } = useRepositoryContext();

  return (
    <Box display="flex" flexDirection="column">
      <Box display="flex">
        <Typography variant="h4" marginRight="10px">
          Reposit√≥rio
        </Typography>
        <Typography variant="h4" fontWeight="500" color="#33568E">
          {currentRepository?.name}
        </Typography>
      </Box>
      <Typography variant="caption" color="gray">
        {currentRepository?.description}
      </Typography>
    </Box>
  );
}

export default Header;
