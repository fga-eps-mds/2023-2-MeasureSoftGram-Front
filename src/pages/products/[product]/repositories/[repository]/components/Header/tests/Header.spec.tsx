import React from 'react';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import { ProductProvider } from '@contexts/ProductProvider';
import { RepositoryProvider } from '@contexts/RepositoryProvider';
import { render } from '@testing-library/react';
import Header from '../Header';

interface Props {
  children: React.ReactNode;
}

const AllTheProviders = ({ children }: Props) => (
  <OrganizationProvider>
    <ProductProvider>
      <RepositoryProvider>{children}</RepositoryProvider>
    </ProductProvider>
  </OrganizationProvider>
);

describe('Header', () => {
  it('should render correctly', () => {
    const { container } = render(<Header />, {
      wrapper: AllTheProviders
    });

    expect(container).toMatchSnapshot();
  });
});
