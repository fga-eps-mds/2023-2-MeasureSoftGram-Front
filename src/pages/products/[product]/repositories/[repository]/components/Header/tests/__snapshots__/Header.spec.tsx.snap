// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-j7qwjs"
  >
    <div
      class="MuiBox-root css-k008qs"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-obdthj-MuiTypography-root"
      >
        Reposit√≥rio
      </h4>
      <h4
        class="MuiTypography-root MuiTypography-h4 css-1yvpk5c-MuiTypography-root"
      />
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption css-qkcnm0-MuiTypography-root"
    />
  </div>
</div>
`;
